Feat: Implement Modal component

YuvrajSinghJadon opened the pull request:
### Changes:
Related to issue #48 
- Added `Modal` component with predefined text variants (`header`, `body`, `footer`).
- Included customization props for `size`, `animationType`, and `dismiss behavior`.
- Ensured accessibility support and font scaling.
- Wrote unit tests covering various use cases.
- Added documentation with usage examples.

---

### Description

This Pull Request introduces the `Modal` component to establish a consistent overlay content system across the application. The component provides predefined structures for header, body, and footer, making it versatile for various use cases like alerts, popups, and confirmation dialogs.

---

### Screenshots

<img width="231" alt="Modal-Demo" src="https://github.com/user-attachments/assets/0340f4b5-411d-49ab-bd84-770d8f45ba88" />


---

### Tests

### Automated Test Cases Added

- **Render Default Variant:** Verifies that the component renders correctly with default props.
- **Apply Variant Styles:** Checks that each size variant (`small`, `medium`, `large`) applies the correct styles.
- **Customization Props Functionality:** Ensures that custom `size`, `animationType`, and `onClose` props are applied correctly.

### Manual Test Cases Run

- **Cross-Platform Rendering:** Verified that the `Modal` component renders correctly on both Android and iOS devices/emulators.
- **Dismiss Behavior:** Tested the close button and background tap to ensure the modal dismisses correctly.
- **Style Overrides:** Applied custom styles via the `style` prop to confirm they override default styles as expected.

---

## Checklist

- [x] All automated tests pass.
- [x] Code has been linted and follows coding standards.
- [x] Documentation is thorough and complete.
- [x] Ready for review.



### Thank you for reviewing this Pull Request. I look forward to your feedback!

jjalan commented on src/components/modal/ README.md:
It seems there is an extra space before `README.md`

jjalan commented on src/components/modal/Modal.tsx:
As mentioned in earlier PR, filename should be case insensitive

jjalan commented on src/components/modal/Modal.tsx:
```diff
@@ -0,0 +1,56 @@
+import { styles } from './modal.styles';
+
+interface ModalProps extends ViewProps {
+  isVisible: boolean;
```
order alphabetically

jjalan replied:
Can we take inspiration from https://baseweb.design/components/modal/ or other populate open source library on how to create abstractions for modal component please?

jjalan commented on src/components/modal/modal.test.tsx:
Test should live outside src folder

YuvrajSinghJadon committed:
Fix case sensitivity issue in modal.tsx

YuvrajSinghJadon committed:
fix lint error

YuvrajSinghJadon committed:
Add Modal component to index exports

YuvrajSinghJadon committed:
Make Modular structure

jjalan commented on src/components/index.ts:
```diff
@@ -1,3 +1,4 @@
 export { default as Brand } from './brand/brand';
 export { default as ErrorFallback } from './error-fallback/error-fallback';
 export { default as OTPInput } from './otp-input/otp-input';
+export { default as Typography } from './modal/modal';
```
Should this be as `default as Modal`?

jjalan commented on src/components/modal/modal.test.tsx:
```diff
@@ -0,0 +1,44 @@
+
+  test('does not render when not visible', () => {
+    const { queryByText } = render(
+      <Modal
```
The usage should be:

```
<Modal>
  <Modal.Header></Modal.Header>
  <Modal.Body></Modal.Body>
  <Modal.Footer></Modal.Footer>
</Modal>
```

Header, Body, and Footer are not really props. They are children. In terms of OOPS, you should think about props like properties of an object.

`BaseWeb` provides a clean interface that can guide us on how to design interfaces of a component. You are welcome to use others. You are welcome to use Tailwind native css to design them.

govind-io commented on src/components/modal/README.md:
```diff
@@ -0,0 +1,36 @@
+- **Close Action:** `onClose` prop to handle the close action.
+- **Content Structure:** `header`, `body`, and `footer` slots for structured content.
+- **Customizable Size:** `size` prop with options ('small', 'medium', 'large').
+- **Animation:** `animationType` prop with options ('none', 'slide', 'fade').
```
Should we also add a prop to allow/disallow **close on click outside behaviour?**

govind-io commented on src/components/modal/README.md:
```diff
@@ -0,0 +1,36 @@
+      <Modal
+        isVisible={isVisible}
+        onClose={() => setIsVisible(false)}
+        header={<Text>Header</Text>}
```
# Feedback on Modal Component Implementation

## Suggested Improvements

The current `Modal` component enforces a predefined structure with `header`, `body`, and `footer` slots, which might limit flexibility in certain use cases. Instead, I suggest making the `Modal` component more generic and unopinionated, allowing developers to pass any content as children without restrictions. 

### Proposed Changes:

1. **Make `Modal` a Generic Container**
   - Instead of predefined slots (`header`, `body`, `footer`), allow developers to structure their content freely within the modal.
   - Maintain core functionalities such as:
     - `isVisible`: Controls visibility.
     - `onClose`: Handles the close action.
     - `size`: Allows different modal sizes (`small`, `medium`, `large`).
     - `animationType`: Defines animation behavior (`none`, `slide`, `fade`).

2. **Create Opinionated Variants Using `Modal`**
   - Introduce specific modal variants (e.g., `ConfirmationModal`) that extend `Modal` while providing default behavior.
   - Example: `ConfirmationModal` should include:
     - Default **"Confirm"** and **"Cancel"** buttons.
     - A **message** (with an option to override or extend the body content).
     - Support for passing additional custom elements if needed.
3. Encourage the Use of Modal Subcomponents
- To promote better structure and consistency, you should create certain children components within Modal, such as:
  - Modal.Header: Contains the title and close button.
  - Modal.Body: Holds the primary content.
  - Modal.Footer: Typically includes action buttons.
  - Modal.Button: A reusable button component inside the modal.

---

## Example Usage

### 1️⃣ Using the Generic `Modal` Component

The `Modal` component allows developers to render any content inside it without predefined structure.

```jsx
import React, { useState } from "react";
import { View, Text, Button } from "react-native";
import Modal from "./Modal"; // Assuming Modal is a generic component

const GenericModalExample = () => {
  const [isVisible, setIsVisible] = useState(false);

  return (
    <View>
      <Button title="Open Modal" onPress={() => setIsVisible(true)} />

      <Modal isVisible={isVisible} onClose={() => setIsVisible(false)}>
        <View style={{ padding: 20 }}>
          <Text style={{ fontSize: 18, fontWeight: "bold" }}>
            Custom Modal Content
          </Text>
          <Text>This modal has no predefined structure.</Text>
          <Button title="Close" onPress={() => setIsVisible(false)} />
        </View>
      </Modal>
    </View>
  );
};

export default GenericModalExample;
```

---

### 2️⃣ Using the `ConfirmationModal` Component

The `ConfirmationModal` extends `Modal` and provides default actions.

```jsx
import React, { useState } from "react";
import { View, Button } from "react-native";
import ConfirmationModal from "./ConfirmationModal"; // Assuming ConfirmationModal extends Modal

const ConfirmationModalExample = () => {
  const [isConfirmVisible, setIsConfirmVisible] = useState(false);

  const handleConfirm = () => {
    console.log("Confirmed!");
    setIsConfirmVisible(false);
  };

  return (
    <View>
      <Button title="Show Confirmation" onPress={() => setIsConfirmVisible(true)} />

      <ConfirmationModal
        isVisible={isConfirmVisible}
        onConfirm={handleConfirm}
        onCancel={() => setIsConfirmVisible(false)}
        message="Are you sure you want to proceed?"
      />
    </View>
  );
};

export default ConfirmationModalExample;
```

---

## **Summary**
- `Modal`: A **generic** and **flexible** modal wrapper.
- `ConfirmationModal`: A **predefined** modal variant using `Modal`, with default actions.

This approach ensures **maximum flexibility** while also providing **reusable patterns** for common use cases. Let me know your thoughts! 🚀

YuvrajSinghJadon replied:
Required changes will be made.

YuvrajSinghJadon committed:
feat: Add modular Modal component with subcomponents

- Introduced `Modal.tsx` as a flexible overlay container.
- Created subcomponents: `ModalHeader`, `ModalBody`, and `ModalFooter` for structured composition.
- Implemented `isVisible` prop to control modal visibility.
- Ensured accessibility support with `onRequestClose` handling.
- Wrote unit tests to verify correct rendering of modal and subcomponents.
- Prepared for future NativeWind integration by keeping styling separate.

YuvrajSinghJadon committed:
refactor:add `ConfirmationModal` Implement PR feedback and improve modularity

- Refactored `Modal` to be a more flexible, generic container.
- Introduced `ConfirmationModal` as an opinionated variant with default actions.
- Added `types.ts` to centralize type definitions and improve maintainability.

govind-io committed:
feat: Update Modal Component

govind-io committed:
feat: update confirmation modal component
